crates/metal-llvm/src/stdlib_priv.rs

```diff
@@ -0,0 +1,82 @@
+use std::{ffi::CString, ptr};
+
+use llvm_sys::{
+    core::LLVMCreateMemoryBufferWithMemoryRangeCopy, ir_reader::LLVMParseIRInContext, prelude::*,
+};
+
+// NOTE: private modules do NOT have submodules.
+/// private modules stand at the core of the stdlib.
+/// They allow us to define metal functionality in the stdlib
+/// for compile time.
+pub trait PrivateSTDLibModule {
+    // private modules usually start with an underscore
+    // But this name does not include that underscore.
+    /// Private STDLib module name.
+    /// i.e. `io`, `cmplib`, etc.
+    fn libname() -> &'static str;
+
+    fn function(
+        ident: metal_ast::Ident,
+        ctx: LLVMContextRef,
+        module: LLVMModuleRef,
+        arguments: Vec<String>,
+    ) -> Option<LLVMValueRef>;
+}
+
+pub struct ExternalLib;
+
+impl PrivateSTDLibModule for ExternalLib {
+    fn libname() -> &'static str {
+        "pext"
+    }
+
+    #[allow(clippy::not_unsafe_ptr_arg_deref)]
```

Comment:
Why was this ignored? As far as i can tell, this lint is triggered correctly as *Ref types are pointers, they are passed as arguments, and are passed down to unsafe functions

Reply:
around line 66 clippy was complaining about ctx possibly being a pointer that's dereferenced

Reply:
The wording is misleading, but the warning it correct. Users of safe Rust can call `.function()` with arbitrary pointer values (including nullptr, because casting and other operations on pointers are safe; only dereferencing pointers is unsafe) and it will happily propagate the pointers down to LLVM without checking them for validity. This means that it's possible to crash the program from safe Rust by calling `.function()` with a bogus pointer value.